pkg/apis/clusterregistry/v1alpha1/types.go

```diff
@@ -167,3 +171,87 @@ type ClusterCondition struct {
+// cluster's healthz endpoint.
+//
+// +k8s:openapi-gen=true
+// +kubebuilder:resource:path=clustercredentials
+// +kubebuilder:subresource:status
+type ClusterCredentials struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterCredentialsSpec   `json:"spec,omitempty"`
+	Status ClusterCredentialsStatus `json:"status,omitempty"`
+}
+
+// ClusterCredentialsSpec defines the desired state of a ClusterCredentials.
+type ClusterCredentialsSpec struct {
+	// ClusterRef is a reference to the Cluster to access.
+	ClusterRef v1.LocalObjectReference `json:"clusterRef,omitempty"`
+
+	// SecretRef is a reference to the secret in the namespace that contains a
+	// kubeconfig with credentials access the referenced cluster. The secret
+	// must have a key named "kubeconfig" that contains a valid kubeconfig.
+	//
+	// TODO: explain required kubeconfig fields
+	// TODO: explain kubeconfig version
+	//
+	// This can be left empty if the cluster allows insecure access.
+	// +optional
+	SecretRef *apiv1.LocalObjectReference `json:"secretRef,omitempty"`
+}
+
+// ClusterCredentialsStatus contains information about the current status of a
+// ClusterCredentials.
+type ClusterCredentialsStatus struct {
+	// Conditions is an array of current conditions.
+	// +optional
+	Conditions []ClusterCredentialsCondition `json:"conditions,omitempty"`
+}
+
+// ClusterCredentialsCondition contains condition information about a ClusterCredentials.
+type ClusterCredentialsCondition struct {
+	// Type is the type of the cluster condition.
+	Type ClusterConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=ClusterCredentialsConditionType"`
+
+	// Status is the status of the condition. One of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
+
+	// LastHeartbeatTime is the last time this condition was updated.
+	// +optional
+	LastHeartbeatTime metav1.Time `json:"lastHeartbeatTime,omitempty" protobuf:"bytes,3,opt,name=lastHeartbeatTime"`
+
+	// Last time the condition was checked.
+	// +optional
+	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty"`
+
+	// LastTransitionTime is the last time the condition changed from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+
+	// Reason is a (brief) reason for the condition's last status change.
+	// +optional
+	Reason string `json:"reason,omitempty" protobuf:"bytes,5,opt,name=reason"`
+
+	// Message is a human-readable message indicating details about the last status change.
+	// +optional
+	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
+}
+
+// ClusterCredentialsConditionType marks the kind of ClusterCredentials condition
+// being reported.
+type ClusterCredentialsConditionType string
```

Comment:
It seems like there may be some overlap with `ClusterConditionType`. Any reason not to use it instead by just adding a new `ClusterCredentialsHealthy` const to it?